compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor.hs

```diff
@@ -413,6 +418,30 @@ checkKinds (GHC.L _ m) = do
         , "This will cause problems when importing this module via data-dependencies."
         ]
 
+-- | Set of package names and module name spaces to emit internal import warnings for
+-- e.g. ("daml-script", "Daml.Script") means that only the "daml-script" package can import "Daml.Script.*.Internal.*" without warning
+-- Module name spaces should be unique and non-overlapping
+warnInternalPackages :: [(LF.PackageName, GHC.ModuleName)]
+warnInternalPackages = fmap (bimap LF.PackageName GHC.mkModuleName)
+  [ ("daml3-script", "Daml.Script")
+  ]
+
+splitModuleName :: GHC.ModuleName -> [String]
+splitModuleName = splitWhen (=='.') . GHC.moduleNameString
+
+-- | Check imports of internal modules of packages in warnInternalPackages given warnings
+checkCustomInternalImports :: GHC.ParsedSource -> Maybe LF.PackageName -> [(GHC.SrcSpan, String)]
+checkCustomInternalImports ps mOwnPkgName = do
+  GHC.L ss GHC.ImportDecl{ideclName=GHC.L _ m} <- GHC.hsmodImports $ GHC.unLoc ps
+  let splitM = splitModuleName m
+  guard $ "Internal" `elem` splitM
+  (pkgName, pkgModuleSpace) <- warnInternalPackages
+  guard $ splitModuleName pkgModuleSpace `isPrefixOf` splitM
```

Comment:
GHC should optimize this away but you could consider binding `splitModuleName pkgModuleSpace` with a let at the beginning of the `do` block or in a where clause.

Reply:
Fair point, i'll do that